modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +330,168 @@ void run(ClusterState state) {
                     e
                 );
             }
+
+            try {
+                maybeExecuteDownsampling(state, dataStream, getTargetIndices(dataStream, indicesToExcludeForRemainingRun));
+            } catch (Exception e) {
+                logger.error(
+                    () -> String.format(
+                        Locale.ROOT,
+                        "Data stream lifecycle failed to execute downsampling for data stream [%s]",
+                        dataStream.getName()
+                    ),
+                    e
+                );
+            }
+        }
+    }
+
+    Set<Index> maybeExecuteDownsampling(ClusterState state, DataStream dataStream, List<Index> targetIndices) {
+        Set<Index> affectedIndices = new HashSet<>();
+        Metadata metadata = state.metadata();
+        for (Index index : targetIndices) {
+            if (dataStream.isIndexManagedByDataStreamLifecycle(index, metadata::index)) {
+                IndexMetadata backingIndex = metadata.index(index);
+                assert backingIndex != null : "the data stream backing indices must exist";
+                List<DataStreamLifecycle.Downsampling.Round> downsamplingRounds = dataStream.getDownsamplingRoundsFor(
+                    index,
+                    metadata::index,
+                    nowSupplier
+                );
+                if (downsamplingRounds.isEmpty()) {
+                    continue;
+                }
+
+                String indexName = index.getName();
+                IndexMetadata.DownsampleTaskStatus backingIndexDownsamplingStatus = INDEX_DOWNSAMPLE_STATUS.get(backingIndex.getSettings());
+                String backingIndexDownsamplingSource = backingIndex.getSettings().get(IndexMetadata.INDEX_DOWNSAMPLE_SOURCE_NAME_KEY);
+
+                // if the current index is not a downsample we want to mark the index as read-only before proceeding with downsampling
+                if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource) == false
+                    && state.blocks().indexBlocked(ClusterBlockLevel.WRITE, indexName) == false) {
+                    // time to downsample, but first we have to mark the index as read-only
+                    affectedIndices.add(index);
+                    AddIndexBlockRequest addIndexBlockRequest = new AddIndexBlockRequest(WRITE, indexName).masterNodeTimeout(
+                        TimeValue.MAX_VALUE
+                    );
+                    transportActionsDeduplicator.executeOnce(
+                        addIndexBlockRequest,
+                        new ErrorRecordingActionListener(indexName, errorStore),
+                        (req, reqListener) -> addIndexBlock(addIndexBlockRequest, reqListener)
+                    );
+                    continue;
```

Comment:
I think the use of `continue` in these few places as part of control flow is confusing and harder to read. We're currently essentially doing something like:

```java
if (hasDownsamplingSource && inState1) {
  doSomethingToGetToState2();
  continue;
}

if (hasDownsamplingSource && inState2) {
  doSomethingToGetToState3();
  continue;
}
```

but I think the mutually exclusive branch execution is cleaner as exclusive if statements (and I would factor out the individual pieces into separate methods so that the control flow reads cleaner:

```java
boolean hasDownsampleSource = Strings.hasText(backingIndexDownsamplingSource);

if (hasDownsampleSource == false && state.blocks().indexBlocked(ClusterBlockLevel.WRITE, indexName) == false) {
    addIndexBlock(indexName);
} else if (hasDownsampleSource && backingIndexDownsamplingStatus.equals(SUCCESS)) {
    replaceIndex(…);
} else {
    executeDownsampling(…);
}
```

I think this would be much easier to read, avoids all the extra `continue` control-flow statements, and avoids this function being _gigantic_.

Reply:
Hm I'm not sure this will work but giving it a try (we have 2 for-loops to tackle). I've extracted a few methods and hopefully it's clearer now